modules/v2x/v2x_proxy/app/v2x_proxy.cc

```diff
@@ -13,147 +13,185 @@
+        !!obu_interface_ && os_interface_->InitFlag() &&
+        obu_interface_->InitFlag());
+  planning_thread_.reset(new std::thread([this]() {
+    while (!exit_) {
+      this->RecvOsPlanning();
+    }
+  }));
   v2x_carstatus_timer_->Start();
-
+  GetRsulistFromFile(FLAGS_rsu_whitelist_name, &rsu_list_);
   init_flag_ = true;
 }
 
+bool V2xProxy::GetRsulistFromFile(const std::string &filename,
+                                  std::set<std::string> *whitelist) {
+  RETURN_VAL_IF_NULL(whitelist, false);
+  std::ifstream input_file(filename);
+  RETURN_VAL_IF(!input_file, false);
+  bool flag = false;
+  std::string strOne;
+  while (getline(input_file, strOne)) {
+    flag = true;
+    whitelist->insert(strOne);
+  }
+  return flag;
+}
+
 bool V2xProxy::InitFlag() { return init_flag_; }
 
-bool V2xProxy::TrafficLightProc(CurrentLaneTrafficLight* msg) {
-  if (!msg->has_gps_x_m() || !msg->has_gps_y_m()) {
-    AERROR << "Error::v2x trafficlight ignore, gps point is null";
-    return false;
-  }
-  apollo::common::PointENU point;
-  point.set_x(msg->gps_x_m());
-  point.set_y(msg->gps_y_m());
-  std::vector<apollo::hdmap::SignalInfoConstPtr> signals;
-  if (hdmap_->GetForwardNearestSignalsOnLane(point, 1000.0, &signals) != 0) {
-    AERROR << "Error::v2x trafficlight ignore, hdmap get no signals";
-    AERROR << "traffic light size : " << signals.size();
-    return false;
-  }
-  for (auto i = signals.begin(); i != signals.end(); i++) {
-    if ((*i)->id().id().empty()) {
-      AERROR << "Error::v2x trafficlight ignore, signals id is empty";
-      return false;
+void V2xProxy::RecvOsPlanning() {
+  auto planning_adc = std::make_shared<::apollo::planning::ADCTrajectory>();
+  auto res_light =
+      std::make_shared<::apollo::perception::TrafficLightDetection>();
+  os_interface_->GetPlanningAdcFromOs(planning_adc);
+  // OK get planning message
+  std::shared_ptr<::v2x::proxy::OSLightT> last_os_light = nullptr;
+  {
+    std::lock_guard<std::mutex> lock(lock_last_os_light_);
+    struct timespec ts = {};
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    if (last_os_light_ == nullptr ||
+        2000LL * 1000 * 1000 <
+            ::v2x::proxy::UtilsDiffTimespec(&ts, &ts_last_os_light_)) {
+      AWARN << "V2X Traffic Light is too old!";
+      last_os_light_ = nullptr;
+    } else {
+      ADEBUG << "V2X Traffic Light is on time.";
+      last_os_light = std::make_shared<::v2x::proxy::OSLightT>();
+      last_os_light->CopyFrom(*last_os_light_);
     }
   }
-  if (signals.size() == 1) {
-    auto single = msg->mutable_single_traffic_light(0);
-    single->set_id(signals[0]->id().id());
-    return true;
-  }
-  auto color = msg->single_traffic_light(0).color();
-  msg->clear_single_traffic_light();
+  // proc planning message
+  bool res_proc_planning_msg = internal_->ProcPlanningMessage(
+      planning_adc.get(), last_os_light.get(), &res_light);
+  RETURN_IF(!res_proc_planning_msg);
```

Comment:
Avoid `RETURN_IF` usage. It hurts readability.
In L4 internal project, `RETURN_IF` usage is forbidden.